4, 5주차: 계획 페이지 세부 기능 작업 및 API 연동

Moderator11 opened the pull request:
# 계획 페이지 세부 기능 작업 및 API 연동

안녕하세요 멘토님, 4, 5주차 산출물에 대한 리뷰 PR을 드립니다.

## 구현 리스트
- Canvas 기능 구현 및 리펙토링, CD 작업
- authContext, axiosInstance 리팩토링

## 개별 작업 소개 - 김성진 
계획 페이지에서 모바일 환경과 데스크탑 환경의 편집 기능의 차별을 두는 기능을 구현했습니다.
또 api 인증/인가 관련 로직을 리팩토링했습니다.

계획페이지 모바일/데스크탑 환경 편집기능 제한
- 계획페이지에서 모바일 환경시에는 편집기능이 제한된다는 UI가 표시되고, 편집기능을 제한하는 기능을 추가했습니다.
- 편집기능의 제한 정도는 flow의 props를 조정하여 필요에 따라 변경할 수 있습니다.
- src/pages/plan/responsive/ 폴더에 관련 로직을 위한 파일을 생성하여 충돌을 피했습니다.
- 기존의 파일 중 변경된 파일로는 App.tsx, Router.tsx, Canvas.tsx가 있습니다. (라우팅 수정)

api 인증/인가 관련 기능 
- 커스텀 이벤트를 WindowEventMap에 전역선언해 핸들러를 정확히 타이핑하고, as EventListener 캐스팅을 제거하여 타입 안전성과 가독성을 높였습니다

아직 머지되지않은 PR이 있어 관련 정보를 간략하게 명시합니다
- axiosInstance의 401에러시의 refresh요청 로직에서 promise를 변수로 관리하던것을 각 요청에 대해 promise를 새롭게 생성해 구독하는 방식으로 변경했습니다.
- 로그인/회원가입 API요청 기능을 구현했습니다. 

## 개별 작업 소개 - 박수민
이번 PR에서는 Canvas 및 계획 페이지 관련 기능 확장과 리팩토링, 배포 자동화 설정을 중심으로 다양한 작업을 진행했습니다.

주요 작업 내역
- React Flow 기반 Canvas 기능 추가
- react-flow 패키지 도입 및 커스텀 노드/타입 정의 추가
- 더미 UI 및 Edge 연결 기능 구현
- Control Bar 구성 및 Edge 연결 기능 구현
- React Flow 제네릭 제약 보일러플레이트 추가

캔버스 및 컴포넌트 구조 개선
- Canvas 컴포넌트 폴더 구조 이동 및 파일 정리
- Props 타입 분리 및 명시적 선언
- Hooks 재명명 및 폴더 이동
- 네임스페이스 제거 및 카테고리 로직 정리 (localStorage 관련 코드 삭제 포함)

CD 작업
- 빌드 및 배포 자동화 및 CD(CI/CD) GitHub Action 세팅 및 관련 설정 추가
- 기타 리팩토링 및 정리(불필요한 import, props, 훅 제거, 타입 경로 오류 수정 등의 빌드 오류 제거 작업)

## 개별 작업 소개 - 남규리
WaypointNode 컴포넌트의 세부 UI를 구현하고 사용자 경험을 개선했습니다.

 - 편집 가능한 필드: 장소, 주소, 메모를 직접 입력하고 수정할 수 있도록 UI를 input과 textarea로 변경했습니다.
 - 동적 크기 조절: useAutosizeInput 훅을 구현하여 텍스트 길이에 따라 입력 필드의 너비가 자동으로 조절되도록 했습니다.
 - 시간 선택 UI: react-datepicker 라이브러리를 도입하여 사용자가 편리하게 출발 및 도착 시간을 선택할 수 있도록 구현했습니다.
 
 카테고리 시스템:
 - 아이콘 클릭 시 카테고리를 선택할 수 있는 커스텀 드롭다운 메뉴를 추가했습니다.
 - 선택된 카테고리에 따라 컴포넌트의 아이콘과 배경색이 동적으로 변경됩니다.
 - 상태 유지: 사용자가 입력한 모든 데이터(장소, 주소, 시간, 메모, 카테고리)는 localStorage에 저장되어, 새로고침해도 내용이 유지됩니다.
 - 코드 분리: 가독성 및 유지보수성 향상을 위해 useAutosizeInput 훅, 카테고리 스타일, CustomTimeInput 컴포넌트를 별도 파일로 분리했습니다.

논의사항:
- Waypoint 노드의 크기가 텍스트의 길이에 따라 자동으로 늘어났다 줄었다 하도록 구현하였는데 텍스트의 길이가 너무 길어질 경우 플랜페이지를 벗어날 수 있다는 단점이 있어 어떻게 노드 크기를 조정하면 좋을지, 텍스트는 어디까지 보이면 좋을지 논의 필요

Moderator11 committed:
chore: react-flow 패키지 추가

Moderator11 committed:
refact: 캔버스 컴포넌트 폴더 이동

Moderator11 committed:
feat: react-flow 커스텀 테스트 노드 추가

Moderator11 committed:
feat: react-flow 커스텀 노드 타입 정의 추가

Moderator11 committed:
feat: Canvas 프로토타입 추가 및 파일 정리

seongjin0320 committed:
refact: 커스텀이벤트 타입 선언으로 캐스팅 제거

- 커스텀 이벤트를 WindowEventMap에 전역선언해 핸들러를 정확히 타이핑하고, as EventListener 캐스팅을 제거하여 타입 안전성과 가독성을 높였습니다

Moderator11 committed:
feat: 노드 더미 UI 구현

seongjin0320 committed:
feat: 계획페이지 캔버스 모바일환경 편집 제한 기능 추가

- src/pages/plan/flow/responsive/ 폴더를 생성해서 반응형 편집제한기능을 구현했습니다
- breakpoints.ts에서 편집가능 최소 화면 너비, 뷰포트 너비 변화 감지 기능을 합니다
- EditGuard.tsx에서 편집 제한 기능이 활성화 될 때의 UI를 출력합니다
- useResponsiveEditing.ts에서 편집 가능 여부를 알려줍니다
- Canvas.tsx에서 ReactFLow의 prop으로 넘겨주는 편집 이벤트를 canEdit일 때만 허용하여 일정 뷰포트 이하가 됐을 때 편집기능을 제한합니다

seongjin0320 committed:
feat: 계획페이지가 늘 화면을 꽉 채우도록 수정

- App.tsx에 BrowserRouter를 위치시키고, 레이아웃 분기는 Router 내부로 이동시켰습니다.
- Router.tsx에 있던 BrowserRouter를 제거하고 path에 따라 MobileLayout, FullBleedLayout을 적용시키도록 했습니다.
- 계획 화면에서만 FullBleedLayout이 적용되도록 했습니다.

Moderator11 committed:
feat: react-device-detect 라이브러리 추가

whyyhyh committed:
feat: waypoint 컴포넌트 수정

whyyhyh committed:
feat: 디폴트 데이터 변경

whyyhyh committed:
refact: 불필요한 주석 삭제

whyyhyh committed:
chore: 패키지 설치(re-resizable)

사용하려다가 실패하여 관련 코드는 빠짐

whyyhyh committed:
feat: 카테고리별 스타일링 추가

whyyhyh committed:
fix: onChange 핸들러 타입 오류 수정

Moderator11 committed:
refact: User-agent 기반 기기 탐지

- react-device-detect 라이브러리 추가
- 폭 기반 기기 탐지 로직 삭제

Moderator11 committed:
feat: Merge pull request #42 from kakao-tech-campus-3rd-step3/feature/plan-responsive-edit

계획페이지 모바일환경시 편집제한 기능 추가, 계획페이지 화면 너비 제한 해제

Moderator11 committed:
Merge branch 'feature/SpacePage' into feature/component-detailed-UI

Moderator11 committed:
fix: 커스텀 prop DOM 전달 차단

- bgColor 프롭이 DOM 트리로 전달되어 React가 인식하지 못하는 문제 해결

Moderator11 committed:
refact: hooks 재명명, 폴더 이동

whyyhyh committed:
refact: 각각의 컴포넌트로 분리

whyyhyh committed:
refactor: 아이콘 컴포넌트 분리

whyyhyh committed:
refactor: props type으로 분리

whyyhyh committed:
refactor: 카테고리 스타일 설정 파일 합치기

whyyhyh committed:
fix: 경로 수정

whyyhyh committed:
refact: interface로 waypoint 세팅 분리

Moderator11 committed:
refact: Props 타입 분리 및 프롭 타입 명시

- useNewPlanForm 훅 프롭 타입 분리 -> UseNewPlanFormProps
- defaultValues 타입 명시

Moderator11 committed:
feat: Merge pull request #48 from kakao-tech-campus-3rd-step3/feature/SpacePage

계획 페이지 canvas 구현

SeonJun-Hwang committed:
Merge pull request #45 from kakao-tech-campus-3rd-step3/refactor/#35FeedBackApply

[박수민] Props 타입 분리 및 프롭 타입 명시

SeonJun-Hwang committed:
Merge pull request #41 from kakao-tech-campus-3rd-step3/refactor/api-auth

[김성진] AuthContext의 이벤트리스너 캐스팅 제거

Moderator11 committed:
Merge branch 'develop' into refactor/LastPR

Moderator11 committed:
feat: Merge pull request #44 from kakao-tech-campus-3rd-step3/refactor/LastPR

[남규리] 컴포넌트 구조 개선 및 타입 안정성 강화

Moderator11 committed:
refact: 네임스페이스 제거

seongjin0320 committed:
refact: 인터셉터 내부 리프레시 promise를 새롭게 생성하여 기다리는방식으로 변경

- promise 변수를 생성하여 여러개의 요청들을 묶었던 방식이 JS의 방식과 맞지 않아 제거했습니다
- promise 를 각 요청마다 새롭게 만들어 구독하여 refresh가 끝나기까지 기다리는 방식으로 변경했습니다

Moderator11 committed:
refact: 카테고리 정리, localStorage 로직 삭제

- localStorage 저장 로직 삭제
- 카테고리, 데이터 인터페이스 통합

Moderator11 committed:
feat: edge 연결 기능 추가

Moderator11 committed:
feat: 웨이포인트, 메모 아이콘 추가

Moderator11 committed:
feat: 컨트롤 바 구성

Moderator11 committed:
fix: react-flow 제네릭 제약 보일러플레이트 추가

Moderator11 committed:
feat: 컨트롤 바 및 Edge 연결 기능 추가

Moderator11 committed:
Merge branch 'develop' into feature/component-detailed-UI

Moderator11 committed:
feat: 캔버스 Edge 연결 및 Control 바 기능 구현, merge #43

- 컴포넌트 세부 디테일 UI 구현
- Edge 연결 기능 추가
- Control 바 추가

Moderator11 committed:
chore: 빈폴더 삭제

Moderator11 committed:
feat: CD Github Action 세팅

Moderator11 committed:
feat: CD 설정, merge #50

CD Github Action 설정

Moderator11 committed:
fix: 미사용 import 구문 삭제

Moderator11 committed:
fix: 미사용 훅 함수 삭제"

Moderator11 committed:
fix: 미사용 props 제거

Moderator11 committed:
fix: PlanPage 컴포넌트 id 사용 로직 추가

Moderator11 committed:
타입 경로 오류 수정

Moderator11 committed:
feat: Merge pull request #51 from kakao-tech-campus-3rd-step3/fix/build

빌드 오류 수정

seongjin0320 committed:
feat: 로그인/회원가입시에 axiosinstance 호출시에 자동으로 accessToken헤더를 붙이지 않도록 변경

- axiosinstance.ts 의 인터셉터에서 기본적으로 API요청에 accessToken헤더를 붙이도록 되어있는데, 회원가입/로그인 시의 POST요청에 member권한이 필요없으므로 accessToken 헤더를 붙이지 않도록 하는 분기로직을 작성했습니다.
- refresh 요청은 axiosinstance.ts내부에서 axios.post를 호출하므로 axiosinstance.ts 바깥에서 refresh로직을 사용할 일이 없을 것 같아 분기로직에서 생략했습니다.

seongjin0320 committed:
feat: 로그인 API 요청 기능 구현

- useLoginForm.ts를 수정하여 로그인 API 요청 기능을 구현했습니다.
- 로그인과 회원가입 API 요청 엔드포인트를 관리하기 편하게 하기 위해 src/api/endpoints.ts에 엔드포인트를 정의하여 상수화 했습니다.

seongjin0320 committed:
feat: 회원가입 API 요청 기능 구현

- 회원가입 API요청 기능을 로그인 API요청과 일관성 있는 방식으로 axiosInstance.post를 통해 구현했습니다
- 회원가입시에 POST요청으로 보내는 데이터에는 비밀번호 확인 필드 값을 제외하고, mbti는 선택 정보이므로 빈값이면 제거해서 보냅니다.

seongjin0320 committed:
fix: 프록시 설정, 회원가입 request 필드가 서버요구와 달랐던 점 수정

- 서버의 CORS가 해결되지 않아 일시적으로 API 기능 확인을 위해 vite설정을 통해 프록시 설정 했습니다
- 서버의 회원가입 요구 필드와 기존의 회원가입 request 필드가 맞지 않아 400에러가 뜨는 것을 해결했습니다
- 회원가입을 통해 계정을 생성한 뒤, 그 계정으로 로그인 시에 로그인이 안되는 문제가 생겨 해결방법 찾는 중입니다

seongjin0320 committed:
chore: 프록시 설정 원래대로 되돌림

seongjin0320 committed:
fix: 회원가입시에 전화번호 데이터가 - 포함으로 그대로 넘겨지도록 수정

seongjin0320 committed:
refact: AuthContext에서 user정보를 다루지 않고, accessToken, refreshToken만을 다루도록 변경

- /login 요청의 반환값에 user가 없어 로그인시 /me 요청을 통해 user값을 받아와 AuthContext에서 토큰정보와 user데이터를 다뤘었지만, user 정보가 필요한 부분에서 /me 요청을 받는 방식으로 사용하기 위해 AuthContext에서 토큰정보만을 다루도록 변경했습니다
- AuthContext에서 토큰 정보만을 다루기 때문에 useloginForm.ts의 onSubmit시에 로그인 성공 후 login()을 이용해 인증 상태를 관리할 때 토큰 값만 이용하도록 변경했습니다
- 확인용 코드를 통해 실제로 토큰 값이 제대로 저장되고, useAuth를 통해 토큰값에 접근할 수 있음을 확인했습니다

seongjin0320 committed:
feat: 회원가입시에 자동 로그인 기능 추가

SeonJun-Hwang commented on src/pages/plan/flow/canvasComponents/Waypoint.ts:
```diff
@@ -0,0 +1,14 @@
+import type { LocationCategory } from '../../utils/Category';
+
+export interface WaypointData extends Record<string, unknown> {
```
Record<string, unknown>를 어떤 목적에서 extends 하시나요?

Moderator11 replied:
reactflow 라이브러리의 Node<T> 타입에서 제네릭 T(T는 커스텀 컴포넌트) 제약 조건 떄문에 캔버스에 사용될  초기화시 아래와 같은 오류가 발생하였습니다.
```
Type 'WaypointData' does not satisfy the constraint 'Record<string, unknown>'.
Index signature for type 'string' is missing in type 'WaypointData'.ts(2344) (alias)
interface WaypointData import WaypointData
```
https://github.com/xyflow/web/issues/486
다른 사람들이 사용하는 라이브러리 코드를 둘러보니 같은 방식으로 커스텀 컴포넌트를 정의하고있는 것 같아 Record<string, unknown>으로 extends 하였습니다.

SeonJun-Hwang replied:
`[key: string]: unknwon`

SeonJun-Hwang commented on src/pages/plan/flow/nodeTypes.ts:
```diff
@@ -0,0 +1,4 @@
+import MemoNode from './nodes/MemoNode';
+import WaypointNode from './nodes/WaypointNode';
+
+export const nodeTypes = { waypoint: WaypointNode, memo: MemoNode };
```
이건 어떤 목적의 변수인가요?

Moderator11 replied:
`src/pages/plan/flow/Canvas.tsx`에서 사용되고있는 ReactFlow 컴포넌트가 key와 Component를 매핑하는 `nodeTypes` props를 받고 있어 해당 props에 들어갈 커스텀 노드 타입을 구성하는 내용을 담은 코드입니다.

SeonJun-Hwang commented on src/pages/plan/flow/nodes/WaypointNode.tsx:
```diff
@@ -0,0 +1,269 @@
+    };
+  }, [isCategorySelectorOpen]);
+
+  const handleDataChange = (field: keyof WaypointData, value: any) => {
+    setData((prev) => ({ ...prev, [field]: value }));
+  };
```
해당 이벤트는 타입을 좀 더 잘 사용할 수 있을거 같아요
any를 사용하지 않는 방향으로 한번 수정해보세요

whyyhyh replied:
제네릭 사용하여 수정하였습니다

SeonJun-Hwang commented on src/pages/plan/flow/responsive/EditGuard.tsx:
```diff
@@ -0,0 +1,41 @@
+
+// 편집 제한 가드 (읽기 전용 안내 배지만 표시, 상호작용은 차단하지 않음)
+type EditGuardProps = {
+  enabled: boolean; // true일 때 읽기 전용 배지 노출
```
enabled는 기본값을 설정하는게 조금 더 좋아보여요

SeonJun-Hwang commented on src/pages/plan/flow/responsive/EditGuard.tsx:
```diff
@@ -0,0 +1,41 @@
+  children: ReactNode;
+};
+
+export default function EditGuard({ enabled, message, children }: EditGuardProps) {
```
EditGuard는 단순히 클릭 이벤트만 막는 느낌이 있어요
실제로 이벤트 자체를 못 일어나게 막는 방법은 없을까요?

Moderator11 replied:
EditGuard는 모바일 환경에서 편집(수정/입력) 동작만 제한하기 위한 컴포넌트입니다.
다만, 캔버스의 이동(panning) 및 확대/축소(zoom) 은 모바일에서도 가능해야 하므로,
전체 이벤트를 차단하지 않고 클릭 기반 편집 이벤트만 제한하도록 구현했습니다.

SeonJun-Hwang commented on src/pages/plan/hooks/useAutosizeInput.ts:
```diff
@@ -0,0 +1,21 @@
+  const ref = useRef<HTMLInputElement>(null);
+  const [width, setWidth] = useState(0);
+
+  useLayoutEffect(() => {
+    if (ref.current) {
+      const span = document.createElement('span');
+      span.style.font = window.getComputedStyle(ref.current).font;
+      span.style.visibility = 'hidden';
+      span.style.position = 'absolute';
+      span.textContent = value || ref.current.placeholder;
+      document.body.appendChild(span);
+      setWidth(span.offsetWidth + 10);
+      document.body.removeChild(span);
+    }
+  }, [value]);
```
useLayoutEffect를 사용한 이유가 있으신가요?

whyyhyh replied:
렌더링된 결과가 화면이 나타나기 전에 실행되도록 하기 위해 사용하였습니다. useEffect를 사용할 경우 너비가 조정되지 않은 상태를 사용자가 볼 수 있기 때문에 useLayoutEffect를 사용하였습니다.

SeonJun-Hwang commented on src/pages/plan/utils/Category.ts:
전체적으로 LocationType을 Key로 사용한 타입이 많은 거 같아요
지금 방식과, 하나 LocationType을 Key로 두고 Value를 확장해 가는 방식 중 어떤게 이번 프로젝트에 더 잘 맞을까요?

whyyhyh replied:
현재 방식이 유지보수시에 번거로울 것 같아 리팩토링하여 수정하였습니다!

SeonJun-Hwang reviewed (changes requested):
한주간 고생 많으셨습니다. 이번 주에는 react-flow를 이용한 개발을 주로 진행하신거 같네요

타입스크립트를 사용하는 만큼, 타입에 대한 안정성을 제공하면 좋겠다는 마음으로 코멘트를 남겼습니다.

전체적으로 개발하실 때, React다운 코드를 작성하고 있는지, 점검하면서 이동하면 좋을 듯 합니다.

Q. WaypointNode에 대해

WaypointNode 크기가 고민이시라면, 크기에 대한 상한을 두고, 스크롤이 생기는 방향으로 만드는 것도 좋아보입니다.

seongjin0320 committed:
refact: PR 피드백 반영

- as const를 통해 AUTH_EXCLUDED_PATH 를 정의했습니다
- 사소한 오타 수정 했습니다
- 피드백 내용중에 회원가입시 전화번호 데이터 양식에 대한 부분이 있었는데 이전 커밋에 적용되었습니다
- useLoginForm.ts와 useRegisterForm.ts의 리디렉션 로직을 navigate가 아닌 라우팅 훅을 사용하도록 했습니다

seongjin0320 committed:
Merge pull request #54 from kakao-tech-campus-3rd-step3/feature/login-register-api

로그인/회원가입 API 요청 기능 구현

Moderator11 committed:
refact: enabled 필드 기본값 설정

Moderator11 committed:
refact: 카테고리 상수 리펙토링

Moderator11 committed:
refact: 데이터 조작 함수 파라미터 타입 구체화

whyyhyh committed:
feat: 제네릭 사용하여 any 타입 제거

whyyhyh committed:
refactor: Category.ts 리팩토링

Moderator11 committed:
fix: .env 삭제 및 gitignore 추가

SeonJun-Hwang committed:
Merge pull request #59 from kakao-tech-campus-3rd-step3/refact/PlanPage

[박수민, 남규리] 피드백 반영

Moderator11 committed:
Merge branch 'develop' into refactor/gyuri

Moderator11 committed:
fix: merge conflict fix, flattenLocationType removal

SeonJun-Hwang committed:
Merge pull request #61 from kakao-tech-campus-3rd-step3/refactor/gyuri

[남규리] PR 피드백 반영
